code/parse/sexp.cpp

```diff
@@ -23298,6 +23567,706 @@ void sexp_set_motion_debris(int node)
+			return false;
+		}
+
+		int modifier_index = 0;
+		for (; modifier_index < MAX_CONTAINER_MODIFIERS; modifier_index++) {
+			if (str_prefix(Container_modifiers[modifier_index].name, text.c_str() + lookHere)) {
+				break;
+			}
+		}
+
+		if (modifier_index == MAX_CONTAINER_MODIFIERS) {
+			Warning(LOCATION,
+				"sexp_replace_container_refs_with_values() found a list container called %s to replace but the modifer "
+				"is not recognised",
+				container.container_name.c_str());
+			return false;
+		}
+
+		int data_index = 0;
+		size_t number_length = 0;
+		SCP_string number_string;
+		auto &list_data = container.list_data;
+		auto list_it = list_data.begin();
+
+		switch (modifier_index) {
+			case SNF_CONTAINER_GET_FIRST:
+				replacement_text = list_data.front();
+				break;
+
+			case SNF_CONTAINER_GET_LAST:
+				replacement_text = list_data.back();
+				break;
+
+			case SNF_CONTAINER_REMOVE_FIRST:
+				replacement_text = list_data.front();
+				list_data.pop_front();
+				break;
+
+			case SNF_CONTAINER_REMOVE_LAST:
+				replacement_text = list_data.back();
+				list_data.pop_back();
+				break;
+
+			case SNF_CONTAINER_GET_RANDOM:
+				data_index = rand_internal(0, (int)list_data.size() -1);
+				replacement_text = *std::next(list_it, data_index);
+				break;
+
+			case SNF_CONTAINER_REMOVE_RANDOM:
+				data_index = rand_internal(0, (int)list_data.size() -1);
+				std::advance(list_it, data_index);
+				replacement_text = *list_it;
+				list_data.erase(list_it);
+				break;
+
+			case SNF_CONTAINER_AT_INDEX:
+				number_string = text.substr(lookHere + strlen(Container_modifiers[SNF_CONTAINER_AT_INDEX].name));
+				Assert(!number_string.empty());
+				data_index = atoi(number_string.c_str());
+				Assert(data_index < (int)list_data.size());
+				replacement_text = *std::next(list_it, data_index);
+
+				// include spaces, in case the FREDder put one between "At" and the index
+				number_length = strspn(number_string.c_str(), " 0123456789");
+				break;
+
+			default:
+				Warning(LOCATION, "sexp_replace_container_refs_with_values() found a container called %s to replace but the modifer is not recognised", container.container_name.c_str());
+				return false;
+		}
+
+		num_chars_to_replace += strlen(Container_modifiers[modifier_index].name);
+
+		// for the "At" modifier, we need to include the chars used by the index
+		if (modifier_index == SNF_CONTAINER_AT_INDEX) {
+			num_chars_to_replace += number_length;
+		}
+
+		// TODO: verify that the char we're about to skip below is a '&'
+	}
+
+	// account for the final '&'
+	++num_chars_to_replace;
+
+	return true;
+}
+
+
+// Karajorma - Based on the same function by Goober for variables.
+/**
+* Replaces container references in a text string with their values
+**/
+bool sexp_replace_container_refs_with_values(SCP_string &text)
+{
+	bool replaced_anything = false;
+	size_t lookHere = 0;
+	int num_replacements = 0;
+
+	while (lookHere < text.length()) {
```

Comment:
This code was originally modeled off of `sexp_replace_variable_names_with_values()`, but since container replacement is more complicated (especially given multidimensionality) and also since `do`-`while` loops are hard to understand IMO, I switched to this version.